zkevm-circuits/src/rlp_circuit_fsm.rs

```diff
@@ -1405,12 +1492,352 @@ impl<F: Field> RlpCircuitConfig<F> {
+        // however, this might change in the future if more nested structures are introduced at same
+        // depth level
+        meta.create_gate(
+            "access list: clearing access_list_idx and storage_key_idx",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::default();
+
+                cb.condition(
+                    meta.query_advice(is_access_list_address_clear, Rotation::cur()),
+                    |cb| {
+                        cb.require_equal(
+                            "al_idx = 0",
+                            meta.query_advice(rlp_table.access_list_idx, Rotation::cur()),
+                            0.expr(),
+                        );
+                    },
+                );
+
+                cb.condition(
+                    meta.query_advice(is_access_list_storage_key_clear, Rotation::cur()),
+                    |cb| {
+                        cb.require_equal(
+                            "sk_idx = 0",
+                            meta.query_advice(rlp_table.storage_key_idx, Rotation::cur()),
+                            0.expr(),
+                        );
+                    },
+                );
+
+                cb.gate(meta.query_fixed(q_enabled, Rotation::cur()))
+            },
+        );
+
+        // Access List Consistency
+        // When no conditions for access list address or storage key changes are present, these idxs
+        // stay the same
+        meta.create_gate(
+            "access list: access_list_idx and storage_key_idx don't change when no conditions present",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::default();
+
+                cb.condition(
+                    and::expr([
+                        not::expr(meta.query_advice(is_new_access_list_address, Rotation::cur())),
+                        not::expr(meta.query_advice(is_access_list_address_clear, Rotation::cur())),
+                    ]),
+                    |cb| {
+                    cb.require_equal(
+                        "al_idx stays the same",
+                        meta.query_advice(rlp_table.access_list_idx, Rotation::prev()),
+                        meta.query_advice(rlp_table.access_list_idx, Rotation::cur()),
+                    );
+                });
+
+                cb.condition(
+                    and::expr([
+                        not::expr(meta.query_advice(is_new_access_list_storage_key, Rotation::cur())),
+                        not::expr(meta.query_advice(is_access_list_storage_key_clear, Rotation::cur())),
+                    ]),
+                    |cb| {
+                    cb.require_equal(
+                        "storage_key_idx stays the same",
+                        meta.query_advice(rlp_table.storage_key_idx, Rotation::prev()),
+                        meta.query_advice(rlp_table.storage_key_idx, Rotation::cur()),
+                    );
```

Comment:
For the storage key, we have a special case where when storage key remains unchanged, its also equal to `0`. So I suggest we change this to a `require_zero` instead of equating `prev` and `cur` row values.

Reply:
Not sure if I understand this suggestion correctly.. a storage key is more than one byte long and will take multiple rows of RLP to decode. In between these rows, the `storage_key_idx` will be consistent but greater than `0`.

Reply:
You are right. My bad, you can ignore this conversation altogether. The current approach is correct.